File: code/nnv/examples/NNV3.0/VideoStar/run_zoomin_4f.py
#!/usr/bin/env python3
"""
VideoStar ZoomIn-4f Verification Script

Optional Python entry point that drives the same MATLAB verification as
``run_zoomin_4f.m`` via the MATLAB engine for Python. The .m file is the
canonical reference implementation; this wrapper is provided for users
who want a Python-native CLI.

Usage:
    python run_zoomin_4f.py [--algorithm {relax,approx,both}] [--num-samples N]

Requirements:
    - MATLAB R2024b with the Python engine installed (`matlab.engine`)
    - NNV toolbox available at <repo>/code/nnv (auto-resolved below)
    - All assets bundled in this folder (models/, data/ZoomIn/, src/vvn/, npy-matlab/)
"""

import argparse
import csv
import datetime
import io
import os
import sys
from typing import Tuple

import matlab.engine


# Paths resolved relative to this file so the script is self-contained.
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
NNV_PATH = os.path.normpath(os.path.join(SCRIPT_DIR, '..', '..', '..'))
VVN_SRC_PATH = os.path.join(SCRIPT_DIR, 'src', 'vvn')
NPY_MATLAB_PATH = os.path.join(SCRIPT_DIR, 'npy-matlab')
RESULTS_DIR = os.path.join(
    SCRIPT_DIR, 'results', datetime.datetime.now().strftime('%y%m%d-%H%M%S')
)


def parse_args():
    parser = argparse.ArgumentParser(
        description='Run ZoomIn-4f verification experiments for VideoStar'
    )
    parser.add_argument(
        '--algorithm',
        type=str,
        choices=['relax', 'approx', 'both'],
        default='relax',
        help='Verification algorithm to use (default: relax)'
    )
    parser.add_argument(
        '--num-samples',
        type=int,
        default=10,
        help='Number of samples to verify (default: 10)'
    )
    parser.add_argument(
        '--timeout',
        type=int,
        default=1800,
        help='Timeout per sample in seconds (default: 1800)'
    )
    return parser.parse_args()


def prepare_engine(nnv_path: str, vvn_src_path: str, npy_matlab_path: str):
    """Start MATLAB engine and add required paths."""
    eng = matlab.engine.start_matlab()
    print('Started MATLAB engine!')

    # Add paths (NNV toolbox + vendored vvn + npy-matlab).
    eng.addpath(eng.genpath(nnv_path))
    eng.addpath(vvn_src_path)
    eng.addpath(npy_matlab_path)

    # verifyvideo.m uses relative paths (`data/ZoomIn/...`, `models/...`),
    # so cd into VideoStar/ where those exist.
    eng.cd(SCRIPT_DIR)

    return eng


def verify(
    ds_type: str,
    sample_len: int,
    ver_algorithm: str,
    eng,
    index: int,
    eps_index: int,
    timeout: int
) -> Tuple[int, float, str]:
    """Run verification on a single sample."""
    if not eng:
        raise Exception(
            'MATLAB Engine was not correctly started. '
            'Please make sure to run `prepare_engine`.'
        )

    # Call MATLAB script to run verification
    future = eng.verifyvideo(
        ds_type,
        sample_len,
        ver_algorithm,
        index,
        eps_index,
        nargout=3,
        background=True,
        stdout=io.StringIO()
    )

    try:
        [res, t, met] = future.result(timeout=float(timeout))
    except matlab.engine.TimeoutError:
        print('  Timeout')
        res = 3
        t = 'timeout'
        met = 'timeout'

    future.cancel()

    return res, t, met


def write_results(output_file: str, sample_num: int, res: int, t, met: str):
    """Append results to CSV file."""
    # Create directory if it doesn't exist
    os.makedirs(os.path.dirname(output_file), exist_ok=True)

    # Write header if file doesn't exist
    if not os.path.exists(output_file):
        with open(output_file, 'w', newline='') as f:
            writer = csv.writer(f)
            writer.writerow(['Sample Number', 'Result', 'Time', 'Method'])

    # Append results
    with open(output_file, 'a', newline='') as f:
        writer = csv.writer(f)
        writer.writerow([sample_num, res, t, met])


def run_zoomin_4f(algorithm: str, num_samples: int, timeout: int):
    """Run ZoomIn-4f verification experiments."""
    # Configuration
    ds_type = 'zoom_in'
    sample_len = 4
    epsilon_indices = [1, 2, 3]  # 1/255, 2/255, 3/255

    # Sample indices to verify (1-indexed for MATLAB)
    sample_indices = list(range(1, num_samples + 1))

    # Output directory: results/<timestamp>/
    output_dir = RESULTS_DIR
    os.makedirs(output_dir, exist_ok=True)

    print("=" * 50)
    print(f"Running ZoomIn-4f verification")
    print("=" * 50)
    print(f"  Algorithm: {algorithm}")
    print(f"  Dataset: {ds_type}")
    print(f"  Sample length: {sample_len} frames")
    print(f"  Number of samples: {num_samples}")
    print(f"  Timeout: {timeout} seconds")
    print(f"  Output directory: {output_dir}")
    print()

    # Start MATLAB engine
    eng = prepare_engine(NNV_PATH, VVN_SRC_PATH, NPY_MATLAB_PATH)

    # Run verification
    for sample_idx, index in enumerate(sample_indices):
        print(f'Sample {index} ({sample_idx + 1}/{num_samples})')

        for eps_index in epsilon_indices:
            output_file = os.path.join(output_dir, f'eps={eps_index}_255.csv')

            # Verify the sample
            res, t, met = verify(
                ds_type, sample_len, algorithm, eng, index, eps_index, timeout
            )

            # Write results
            write_results(output_file, index, res, t, met)

            # Print result
            if isinstance(t, (int, float)):
                print(f'  eps={eps_index}/255: Result={res}, Time={t:.2f}s')
            else:
                print(f'  eps={eps_index}/255: Result={res}, Time={t}')

    # Close MATLAB engine
    eng.quit()

    print()
    print("=" * 50)
    print("Verification complete.")
    print(f"Results saved to: {output_dir}")
    print("=" * 50)


def main():
    args = parse_args()

    # Determine which algorithms to run
    algorithms = []
    if args.algorithm in ['relax', 'both']:
        algorithms.append('relax')
    if args.algorithm in ['approx', 'both']:
        algorithms.append('approx')

    for algorithm in algorithms:
        run_zoomin_4f(algorithm, args.num_samples, args.timeout)


if __name__ == "__main__":
    main()

File: code/nnv/examples/NNV3.0/VideoStar/run_zoomin_4f.m
%% VideoStar ZoomIn-4f - Main Runner Script
% Runs a subset of the ZoomIn-4f video verification experiments.
%
% USAGE:
%   matlab -batch "cd code/nnv/examples/NNV3.0/VideoStar; run_zoomin_4f"
%
% OUTPUTS (under VideoStar/results/<timestamp>/):
%   - eps=<i>_255.csv: per-epsilon CSV with sample, result, time, method
%
% REQUIREMENTS:
%   - NNV toolbox available at <repo>/code/nnv (resolved automatically below)
%   - Bundled ONNX model in VideoStar/models/zoomin_4f.onnx
%   - Bundled data in VideoStar/data/ZoomIn/*.npy
%   - Bundled npy-matlab and vvn src in this folder

%% ================== CONFIGURATION ==================
% Defaults below are applied only for fields the caller has not already
% set in a pre-populated `config` struct (e.g., from run_all.sh --smoke).

if ~exist('config', 'var'); config = struct(); end
scriptDir = fileparts(mfilename('fullpath'));

% Path to NNV root folder (three levels up from VideoStar/).
if ~isfield(config, 'nnvDir');        config.nnvDir       = fullfile(scriptDir, '..', '..', '..'); end

% Bundled assets next to this script.
if ~isfield(config, 'modelsDir');     config.modelsDir    = fullfile(scriptDir, 'models'); end
if ~isfield(config, 'dataDir');       config.dataDir      = fullfile(scriptDir, 'data'); end
if ~isfield(config, 'npyMatlabDir');  config.npyMatlabDir = fullfile(scriptDir, 'npy-matlab'); end
if ~isfield(config, 'vvnSrcDir');     config.vvnSrcDir    = fullfile(scriptDir, 'src', 'vvn'); end

% Timestamped output: results/<yymmdd-HHmmss>/
if ~isfield(config, 'outputDir')
    ts = char(datetime('now', 'Format', 'yyMMdd-HHmmss'));
    config.outputDir = fullfile(scriptDir, 'results', ts);
end

% Verification settings
if ~isfield(config, 'dsType');        config.dsType       = 'zoom_in'; end       % 'zoom_in' or 'zoom_out'
if ~isfield(config, 'sampleLen');     config.sampleLen    = 4; end               % 4, 8, or 16
if ~isfield(config, 'verAlgorithm');  config.verAlgorithm = 'relax'; end         % 'relax' or 'approx'
if ~isfield(config, 'numClasses');    config.numClasses   = 10; end
if ~isfield(config, 'epsilon');       config.epsilon      = [1/255; 2/255; 3/255]; end
if ~isfield(config, 'timeout');       config.timeout      = 1800; end
if ~isfield(config, 'sampleIndices'); config.sampleIndices = 1:10; end

%% ================== END CONFIGURATION ==================

%% Initialize NNV
disp("======= VideoStar ZoomIn-4f Verification ==========");
disp(" ");
disp("Initializing NNV toolbox...");

% Check if NNV directory exists
if ~exist(config.nnvDir, 'dir')
    error("NNV directory not found: %s", config.nnvDir);
end

% Add NNV to path
addpath(genpath(config.nnvDir));
disp("NNV paths added successfully.");

% GPU forward-compat (Blackwell / RTX 5090 under MATLAB R2024b).
% No-op on hosts without an NVIDIA GPU.
try
    parallel.gpu.enableCUDAForwardCompatibility(true);
catch
end

% Add npy-matlab to path for reading .npy files
if ~exist(config.npyMatlabDir, 'dir')
    error("npy-matlab directory not found: %s", config.npyMatlabDir);
end
addpath(config.npyMatlabDir);
disp("npy-matlab paths added successfully.");

% Add bundled vvn src directory to path (for verifyvideo.m)
addpath(config.vvnSrcDir);
disp("vvn verification functions added to path.");

disp(" ");

%% Validate Configuration
disp("Validating configuration...");

% Check if models directory exists
if ~exist(config.modelsDir, 'dir')
    error("Models directory not found: %s", config.modelsDir);
end

% Check if data directory exists
if ~exist(config.dataDir, 'dir')
    error("Data directory not found: %s", config.dataDir);
end

% Check if model exists
modelName = sprintf("zoomin_%df.onnx", config.sampleLen);
modelPath = fullfile(config.modelsDir, modelName);
if ~exist(modelPath, 'file')
    error("Model not found: %s", modelPath);
end
disp("  Found model: " + modelName);

% Create output directory if it doesn't exist
if ~exist(config.outputDir, 'dir')
    mkdir(config.outputDir);
    disp("  Created output directory: " + config.outputDir);
else
    disp("  Output directory: " + config.outputDir);
end

disp(" ");
disp("Configuration validated successfully.");
disp(" ");

%% Change to VideoStar directory so verifyvideo.m's relative paths
% (`data/ZoomIn/...`, `models/...`) resolve against this folder.
currentDir = pwd;
cd(scriptDir);
disp("Changed working directory to VideoStar for data loading.");
disp(" ");

%% Run Verification
disp("======= Running ZoomIn-4f Verification ==========");
disp(" ");
disp(sprintf("Dataset: %s", config.dsType));
disp(sprintf("Sample Length: %d frames", config.sampleLen));
disp(sprintf("Verification Algorithm: %s", config.verAlgorithm));
disp(sprintf("Number of samples: %d", length(config.sampleIndices)));
disp(sprintf("Epsilon values: [1/255, 2/255, 3/255]"));
disp(sprintf("Timeout: %d seconds", config.timeout));
disp(" ");

% Initialize results storage
numSamples = length(config.sampleIndices);
numEpsilons = length(config.epsilon);
results = cell(numSamples, numEpsilons);
times = cell(numSamples, numEpsilons);
methods = cell(numSamples, numEpsilons);

% Run verification for each sample and epsilon
for sampleIdx = 1:numSamples
    sampleNum = config.sampleIndices(sampleIdx);
    disp(sprintf("Processing sample %d (%d/%d)...", sampleNum, sampleIdx, numSamples));

    for epsIdx = 1:numEpsilons
        eps = config.epsilon(epsIdx);
        disp(sprintf("  Epsilon = %d/255...", epsIdx));

        try
            % Call the verification function from FORMALISE2025
            [res, t, met] = verifyvideo(config.dsType, config.sampleLen, ...
                                        config.verAlgorithm, sampleNum, epsIdx);

            results{sampleIdx, epsIdx} = res;
            times{sampleIdx, epsIdx} = t;
            methods{sampleIdx, epsIdx} = met;

            disp(sprintf("    Result: %d, Time: %.2f s", res, t));

        catch ME
            disp(sprintf("    Error: %s", ME.message));
            results{sampleIdx, epsIdx} = -1;
            times{sampleIdx, epsIdx} = -1;
            methods{sampleIdx, epsIdx} = ME.message;
        end
    end
end

%% Save Results
disp(" ");
disp("======= Saving Results ==========");

% Save results for each epsilon value
for epsIdx = 1:numEpsilons
    outputFile = fullfile(config.outputDir, sprintf('eps=%d_255.csv', epsIdx));

    % Open file and write header
    fid = fopen(outputFile, 'w');
    fprintf(fid, 'Sample Number,Result,Time,Method\n');

    % Write results
    for sampleIdx = 1:numSamples
        sampleNum = config.sampleIndices(sampleIdx);
        res = results{sampleIdx, epsIdx};
        t = times{sampleIdx, epsIdx};
        met = methods{sampleIdx, epsIdx};

        if isnumeric(t)
            fprintf(fid, '%d,%d,%.6f,%s\n', sampleNum, res, t, met);
        else
            fprintf(fid, '%d,%d,%s,%s\n', sampleNum, res, t, met);
        end
    end

    fclose(fid);
    disp(sprintf("  Saved: %s", outputFile));
end

%% Display Summary
disp(" ");
disp("======= Verification Summary ==========");
disp(" ");

for epsIdx = 1:numEpsilons
    verified = 0;
    unknown = 0;
    violated = 0;
    timeout = 0;
    totalTime = 0;
    validTimes = 0;

    for sampleIdx = 1:numSamples
        res = results{sampleIdx, epsIdx};
        t = times{sampleIdx, epsIdx};

        if res == 1
            verified = verified + 1;
        elseif res == 0
            violated = violated + 1;
        elseif res == 2
            unknown = unknown + 1;
        elseif res == 3
            timeout = timeout + 1;
        end

        if isnumeric(t) && t > 0
            totalTime = totalTime + t;
            validTimes = validTimes + 1;
        end
    end

    avgTime = totalTime / max(validTimes, 1);

    disp(sprintf("Epsilon = %d/255:", epsIdx));
    disp(sprintf("  Verified (robust): %d", verified));
    disp(sprintf("  Violated: %d", violated));
    disp(sprintf("  Unknown: %d", unknown));
    disp(sprintf("  Timeout: %d", timeout));
    disp(sprintf("  Average time: %.2f s", avgTime));
    disp(" ");
end

%% Cleanup
cd(currentDir);

disp("======= VideoStar ZoomIn-4f Complete ==========");
disp(" ");
disp(sprintf("Results saved to: %s", config.outputDir));
